typing/typecore.ml

```diff
@@ -3229,14 +3219,20 @@ let rec approx_type env sty =
         let tyl = List.map (approx_type env) ctl in
         newconstr path tyl
       end
-  | Ptyp_poly (_, sty) ->
-      approx_type env sty
   | _ -> newvar ()
 
 let type_pattern_approx env spat ty_expected =
   match spat.ppat_desc with
   | Ppat_constraint (_, sty) ->
-      let inferred_ty = approx_type env sty in
+      let inferred_ty =
+        match sty with
+        | {ptyp_desc=Ptyp_poly _} ->
+          let inferred_ty =
+            Typetexp.transl_simple_type env ~closed:false sty
```

Comment:
This lines makes a difference between the two following examples :
```ocaml
let rec f () = g (module Map.Make(Int))
and g (m : (module Map.S)) = ();;
```
and
```ocaml
let rec f () = g (module Map.Make(Int))
and g (m : ('a. module Map.S)) = ();;
```
Where the second one is well-typed, while the was rejected by the type-checker.